design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,319 @@
+
+- Venafi: Makes use of annotations, and so these annotations should also be
+    present on `CertificateSigningRequests`.
+
+- ACME: Makes use of annotations, and so these annotations should also be
+    present on `CertificateSigningRequests`.
+
+- ACME: The ACME controller creates sub-resources (`Orders`). Since
+    `CertificateSigningRequests` are cluster scoped resources, when referencing
+    a `ClusterIssuer` we should create `Orders` in the `Cluster Resource
+    Namespace` (default `cert-manager`), else the namespace of the referenced
+    `Issuer`.
+
+
+### Conditions
+
+The `CertificateSigningRequest` has [well-known condition
+types](https://github.com/kubernetes/api/blob/4a626d306b987a4096cf0784ec01af1be2f6d67f/certificates/v1/types.go#L222)
+of `Approved`, `Denied`, and `Failed`. All `CertificateSigningRequest` signer
+controllers should not begin computation of the request until the resource has
+the `Approved` condition set to `True`.
+
+In the absence of further policy, the `Approved` condition is always set t
+`True`, _unless_, the `CertificateSigningRequest` references a namespaced
+`Issuer` and the requester does not have create permissions for
+`certificates.k8s.io/CertificateSigningRequests` in that namespace. In the case
+of this namespace validation failing, the `CertificateSigningRequest` `Denied`
+condition should be set to `True`, and no signer controllers should process the
+request.
+
+If a signer fails when processing a `CertificateSigningRequest`, the well-known
+`Failed` condition should be set to `True`. No signer controllers should process
+the request again.
+
+If the referenced issuer resource doesn't exist or is not in a ready state, the
+corresponding `CertificateSigningRequest` controller should set the `Pending`
+condition to `True`.
+
+When the request has been successfully signed, the [cert-manager `Ready`
+condition](https://github.com/jetstack/cert-manager/blob/969b678f330c68a6429b7a71b271761c59651a85/pkg/apis/certmanager/v1/types_certificaterequest.go#L139)
+should be set to `True` on the resource, to match the same behaviour as
+`CertificateRequests`.
+
+
+### API Changes
+
+* The `cert-manager.io/duration` annotation on `CertificateSigningRequests`
+    denotes the requested duration as a Go duration string.
+
+* The `cert-manager.io/ca` annotation on `CertificateSigningRequests` contains
+    the returned base 64 PEM encoded CA certificate from the signer.
+
+* The requester of a `CertificateSigningRequest` must have `create` permissions
+    for `certificates.k8s.io/CertificateSigningRequests` in the same namespace
+    the referenced issuer resides.
+
+* The [Approval
+    Condition](https://github.com/kubernetes/api/blob/4a626d306b987a4096cf0784ec01af1be2f6d67f/certificates/v1/types.go#L222)
+    will be replicated on the `CertificateRequest` resource. This condition will
+    _always_ be set to true, in the absence of policy. All `CertificateRequest`
```

Comment:
This sounds scary.

Reply:
I agree... An introduction of a policy engine is pinned for an alpha release in 1.3. This however matches the same behaviour that we have today.